be/src/exec/pipeline/scan_operator.cpp

```diff
@@ -8,177 +8,134 @@
-        auto num_scan_operators = 1 + state->exec_env()->increment_num_scan_operators(1);
-        if (num_scan_operators > _io_threads->get_queue_capacity()) {
-            state->exec_env()->decrement_num_scan_operators(1);
-            return Status::TooManyTasks(
-                    strings::Substitute("num_scan_operators exceeds queue capacity($0) of pipeline_pool_thread",
-                                        _io_threads->get_queue_capacity()));
-        }
+    DCHECK(_io_threads != nullptr);
+    auto num_scan_operators = 1 + state->exec_env()->increment_num_scan_operators(1);
+    if (num_scan_operators > _io_threads->get_queue_capacity()) {
+        state->exec_env()->decrement_num_scan_operators(1);
+        return Status::TooManyTasks(
+                strings::Substitute("num_scan_operators exceeds queue capacity($0) of pipeline_pool_thread",
+                                    _io_threads->get_queue_capacity()));
     }
     _pickup_morsel(state);
     return Status::OK();
 }
 
 Status ScanOperator::close(RuntimeState* state) {
-    if (_io_threads != nullptr) {
-        state->exec_env()->decrement_num_scan_operators(1);
-    }
+    state->exec_env()->decrement_num_scan_operators(1);
     if (_chunk_source) {
+        DCHECK(_io_task_state.load(std::memory_order_acquire) == IOTaskState::INACTIVE);
         _chunk_source->close(state);
     }
     Operator::close(state);
     return Status::OK();
 }
 
-bool ScanOperator::_has_output_blocking() const {
-    DCHECK(_io_threads == nullptr);
-    return _chunk_source->has_next_chunk();
-}
-
-bool ScanOperator::_has_output_nonblocking() const {
-    DCHECK(_io_threads != nullptr);
-    // EOS has arrived
+bool ScanOperator::has_output() const {
     if (_is_finished) {
         return false;
     }
-    DCHECK(_pending_chunk_source_future.has_value());
-    // submitted io task has not completed yet
-    if (_pending_chunk_source_future.value().wait_for(std::chrono::seconds::zero()) != std::future_status::ready) {
+
+    if (!_has_next_morsel) {
         return false;
     }
-    // submitted io task has already completed
-    return true;
-}
 
-bool ScanOperator::has_output() const {
-    if (_io_threads != nullptr) {
-        return _has_output_nonblocking();
-    } else {
-        return _has_output_blocking();
+    DCHECK(_chunk_source != nullptr);
+
+    // Still have cached chunks
+    if (_chunk_source->has_output()) {
+        return true;
+    }
+
+    // io task is busy reading chunks, so we just wait
+    if (_io_task_state.load(std::memory_order_acquire) != IOTaskState::INACTIVE) {
+        return false;
     }
+
+    // Here are two situation
+    // 1. Cache is empty out and morsel is not eof, _trigger_next_scan is required
+    // 2. Cache is empty and morsel is eof, _pickup_morsel is required
+    // Either of which needs to be triggered in pull_chunk, so we return true here
+    return true;
 }
 
 bool ScanOperator::pending_finish() {
```

Comment:
that pending_finish always return false is invalid. if pending io tasks just stay in task queue or grabbed by io threads soon after it is poped from the task queue, in such scenarios, A prematurely finished ScanOperator should wait for pending io tasks' completion.

Reply:
You may misunderstanding the logic here, it means that the chunk cache is empty, and one of the actions should be triggerred(list as follows), and returning true here is to trigger the call of pull_chunk which will trigger these two actions
1. scan next batch of chunks
2. pick up next morsel

But after one of the action is triggerred, it can switch into pending state waiting for the incoming chunk